start

```diff
@@ -225,6 +236,10 @@ function init_db() {
 	chown postgres:postgres $PGDATA
 	chmod 0700 $PGDATA
 
+	# In some environments this directory is missing.
```

Comment:
I am not sure why you suddenly need this. Maybe check with ops they may have a better idea?

Reply:
Wasn't able to get input from ops. It looks to be specific to the DigitalOcean App platform. When building the image on Heroku the dir shows up (we can't use Heroku though because they block root permissions). It also shows up when I build using the DigitalOcean VMs. 

The DO App platform is relatively new, so I'm guessing it's related to that, I posted to their forum [here](https://www.digitalocean.com/community/questions/docker-app-missing-var-run-postgresql-folder) to see.

Reply:
@MonsieurNicolas thoughts on merging this in as a workaround? (imo pretty low risk) I changed the comment here to specify it's needed for DO in case we use something else or the error is fixed on their end:
https://github.com/stellar/docker-stellar-core-horizon/pull/276/commits/325e4e9f548c799ebbdebfb185f4e005c15bdd76

alternatively we could hold off on merging this into main until that's resolved, and add the quickdeploy button to the payment channels branch only, as that was the main goal for this ticket anyways.

Reply:
Reason I was asking is that it kinda looks like a hack and it's unclear to me that this is a digital ocean only problem (could be a pg vs ubuntu related issue):
* we already set all PG related variables in the script
* we're supposed to do all setup work based off those environment variables
* we're supposed to run all pg commands so that we only depend on those environment variables
* if there is a hidden dependency on some other folder, this may point to a bug in the script where we don't actually use one of the folders like we think we are (the script is supposed to create a complete standalone database completely isolated from the default one). For example: we may not use the right database instance which can cause future us to waste a bunch of time in the future debugging.

Reply:
I'm suspicious this is a user permissions or UID issue, where UIDs aren't aligning at runtime. This isn't something that's guaranteed at runtime in every docker environment. That could cause the service to fail to create the directory.

Reply:
to me it doesn't look to be a bug in our code. I'm fairly certain the issue is in how DO builds the image and then deploys it. Based off going through the build logs and they look the same for DO vs local. As well as being able to manually build the image and start a container on DO's App platform and it working fine.

But I agree creating the directory is kind of hacky, and we're changing our code for a 3rd party bug. I'm going to put this PR back into draft, and look into workarounds.

I already found one where we can put a pre-built image in a hub and then tell DO to pull from there instead of building it, which works if I do it manually. But there's a [different bug](https://www.digitalocean.com/community/questions/app-platform-deploying-with-image-from-docker_hub-not-loading) of using the deploy button that I posted to DO that hopefully they can resolve soon.

Reply:
Correct me if I'm wrong here, but I think we still don't know what error is occurring. We don't have a log that is telling us what's failing to start or why. We do have logs showing we can't connect to postgresql, but I think that is a symptom.

Reply:
Yup that's correct. The core problem isn't clear, the build logs look fine with no error. But during deployment the directory is missing. It looks to be a bug on DO's side, the UID issue you pointed out could be the culprit

Reply:
I did some testing.

When running I see the postgres start fail and the logs show it fails to create the lock file.
```
2021-09-28 22:34:42 UTC [49-4] FATAL:  could not create lock file "/var/run/postgresql/.s.PGSQL.5432.lock": No such file or directory
```
Indeed it appears the directory doesn't exist.

Postgresql creates the `/var/run/postgresql` directory during installation, it is part of the startup scripts that get run when it is installed with `apt install`. Debugging shows that Digital Ocean's containers during build have `/run`, but do not have `/var/run`. That's pretty odd given the `ubuntu:20.04` image has a `/var/run` symbolic link to `/run`, but it is possible for their container runtime to be doing something different. The `/var/run` exists during container run but not during the build. This is not extraordinary given that `/var/run` and `/run` are typically tmpfs volumes, which means they are transient and lost during reboot, and so the fact that the `/var/run/postgresql` directory doesn't exist later on should not be a problem. i.e. We shouldn't rely on the that folder already existing.

A little research shows that postgresql when booted correctly using it's init scripts on Ubuntu, or using `pg_ctlcluster` which is the correct way to start postgres on Ubuntu outside of init scripts, both will create the `/var/run/postgresql` as part of its startup process. It's important it does this because there's no guarantee `/var/run/postgresql` will exist after a reboot. However, we have our own non-standard custom approach to booting postgresql in the `start` script as part of the setup. We boot postgres using our own approach, do some things with it, then we shut it down and start it again with `supervisord`. So I think the problem is we run postgrsql manually ourselves and so we are responsible for creating `/var/run/postgresql` if that's the location we plan to place the unix socket. Even when we run postgres inside supervisord we're not starting it using an init script, we're doing it directly.

If you're curious and want to see for ourself, if you have a look at the default `/etc/init.d/postgresql` installed in Ubuntu you'll see when the `start` command is passed it calls a function `create_socket_directory`. Or, if you look at the source for pg_ctlcluster it does so too: https://github.com/credativ/postgresql-common/blob/master/pg_ctlcluster#L447.

So the fact it works today is a bit of a fluke of our local docker deployments. Our approach is to start postgres manually and so we are responsible for making sure the socket directory exists.

We could refactor the startup script and our supervisord config to use init scripts or pg_ctlcluster to start and stop postgres as required and either should take care of creating the directory, but I think a refactor like that is out-of-scope of this PR.

In the end this is a bug in our start script that is only apparent in some environments. The start script's approach is to start postgres manually, so it must be responsible for making sure `/var/run/postgresql` exists. I think we should move ahead with `mkdir -p /var/run/postgresql` and change the comment to something like:
```suggestion
# Create /var/run/postgresql because we aren't using supervisord to
# start postgres it is our responsibility to make sure the directory
# exists for where the process files and unix socket will live.
```

Note that we could of course choose to put the socket in some other folder, such as in the `/opt/stellar` directory path, but that doesn't remove the need to create the directory, and it will be inconvenient if the socket doesn't live at the default path.

Reply:
@MonsieurNicolas looks like I was wrong it is a bug in our script based off of Leigh's findings above. In which case do you think it's reasonable to merge this in?

Reply:
>wasn't trivial for me to figure out how to change it to look somewhere else

The way to make it look somewhere else is to set the `PGHOST` env var, e.g. `PGHOST=/opt/stellar/postgresql/run`. However, you need to make sure it is set for any application that will access the postgres connection. It's trivial to add the env var to the start script and it will apply to things in there. But the moment you `sudo -u postgres` the env will be lost, and if you docker exec into the container it also won't be set, so the env var would need setting in the `/etc/environment` so it applies to all users and shells. In my opinion this would be too inconvenient, so I think we should keep the socket directory where it is.